### PR TITLE
Deep clone cache to avoid transactions from affecting central cache.

### PR DESCRIPTION
**Pull request by fppt**
The need for this feature is best exhibited by the test attached in this PR.
It's to avoid a transaction from affecting the centrally cached ontology. The only way to do this is to deep clone the cached ontology at the start of a transaction.

Please review this PR very carefully. As with other caching PRs this has the potential to cause a lot of harm.

### CODE REVIEW

**Comment by fppt** [comment]
please retest this please

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java [hunk 1]
```diff
@@ -285,15 +288,105 @@ private EdgeImpl addEdge(Concept from, Concept to, Schema.EdgeLabel type){
...
+        for (TypeImpl<?, ?> type : getCloneCache().values()) {
+            TypeImpl centralType = cachedOntologySnapshot.get(type.getName());
+
+            if(type.isRelationType()){
+                ((RelationTypeImpl) type).copyCachedConcepts((RelationTypeImpl) centralType);
+            } else if(type.isRoleType()){
+                ((RoleTypeImpl) type).copyCachedConcepts((RoleTypeImpl) centralType);
+            } else {
+                type.copyCachedConcepts(centralType);
+            }
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: Why can't we just call `type.copyCachedConcepts`?

Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: We need to pass in the type to be copy the sets from and the reason we need the casting is because it will default to the `TypeImpl` copier even if it is a `RelationTypeImpl`. I hate this chunk of code as well but by not casting the wrong copier is called.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java [hunk 2]
```diff
@@ -285,15 +288,105 @@ private EdgeImpl addEdge(Concept from, Concept to, Schema.EdgeLabel type){
...
+            clonedType = new RelationTypeImpl((RelationTypeImpl) type);
+        } else if(type.isEntityType()) {
+            clonedType = new EntityTypeImpl((EntityTypeImpl) type);
+        } else if(type.isRuleType()) {
+            clonedType = new RuleTypeImpl((RuleTypeImpl) type);
+        } else if(type.isResourceType()) {
+            clonedType = new ResourceTypeImpl((ResourceTypeImpl) type);
+        } else if(type.isType()) {
+            clonedType = new TypeImpl((TypeImpl) type);
+        }
```
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: Ahhh I agree. I can make that change.

Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: Since I am travelling would you mind letting me do this in another PR since its more of a code cleanup than a bug or issue? 

I want to get this is quickly so we can find issues with it faster.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java (hunk 1 above)
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: For completeness' sake, here's why we can't use the `TypeImpl` generic `T` as the parameter: "because the cache stores everything as a *Impl you still need to do some odd casting which results in methods being bypassed again."

I would suggest that we just have a method that accepts `TypeImpl`, overridden by `RelationTypeImpl` and `RoleTypeImpl`, which will handle their logic (as well as the dirty casts necessary).

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java (hunk 2 above)
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: Yeah, that's fine.
